### PR TITLE
OGM-1462 Support composite indexes in Neo4j

### PR DESCRIPTION
**Pull request by ntviet18**
https://hibernate.atlassian.net/browse/OGM-1462

Hi @DavideD , @gsmet , could you please review?

### CODE REVIEW

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/CompositeIndexDetails.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.List;
+
+import org.hibernate.ogm.datastore.neo4j.query.parsing.cypherdsl.impl.CypherDSL;
+
+import org.neo4j.graphdb.Label;
+
+/**
+ * @author The Viet Nguyen &amp;ntviet18@gmail.com&amp;
+ */
+public class CompositeIndexDetails {
```
Comment (gsmet) [inline neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/CompositeIndexDetails.java]: AFAICS, it also create mono-column indexes. I would call it `NeojIndexSpec` to be consistent with `MongoDBIndexSpec`.

(I'll let @DavideD do a proper review as he's more proficient in Neo4j than me)

**Comment by gsmet** [comment]
@DavideD I suspect the Neo4j version we have on CI is too old for this.

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/CompositeIndexDetails.java (hunk 1 above)
Comment (DavideD) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/CompositeIndexDetails.java]: I think it's consistent with the existing classes in Neo4j, I will keep this in mind and change it later for the classes involved.

Comment (ntviet18) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/CompositeIndexDetails.java]: yes, it's aligned with `UniqueConstraintDetails` in Neo4j module

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/BaseNeo4jSchemaDefiner.java [hunk 2]
```diff
@@ -23,20 +25,25 @@
...
-import java.lang.invoke.MethodHandles;
 import org.hibernate.ogm.datastore.spi.BaseSchemaDefiner;
 import org.hibernate.ogm.datastore.spi.DatastoreProvider;
 import org.hibernate.ogm.model.key.spi.IdSourceKeyMetadata;
 import org.hibernate.service.spi.ServiceRegistryImplementor;
 import org.hibernate.tool.hbm2ddl.UniqueConstraintSchemaUpdateStrategy;
+
 import org.jboss.logging.Logger.Level;
+
 import org.neo4j.graphdb.Label;
```
[the remark this answers is not in the document]
Comment (ntviet18) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/BaseNeo4jSchemaDefiner.java]: sorry, I forgot to disable auto organize imports

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/CompositeIndexDetails.java (hunk 1 above)
Comment (DavideD) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/CompositeIndexDetails.java]: I just realized that this is different than unique constraints. Le't call it `Neo4jIndexSpec` as suggested by @gsmet. This way we also highlight that they are two different concepts

File: neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/mapping/CompositeIndexTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+		}
+
+		public void setLastname(String lastname) {
+			this.lastname = lastname;
+		}
+	}
+
+	@Test
+	public void test() {
+		// TODO: create test plans
```
Comment (DavideD) [inline neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/mapping/CompositeIndexTest.java]: In MongoDB there are already a lot of test cases for this issue that you can copy and adapt.  MongoDB is more powerful in term of indexes you will need to pick the one that make sense for Neo4j.

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/embedded/impl/EmbeddedNeo4jSchemaDefiner.java [hunk 4]
```diff
@@ -72,6 +97,33 @@ protected void createUniqueConstraintsIfMissing(DatastoreProvider provider, List
 		}
 	}
 
+	private void createCompositeIndex(GraphDatabaseService neo4jDb, CompositeIndexDetails index) {
```
Comment (gsmet) [inline neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/embedded/impl/EmbeddedNeo4jSchemaDefiner.java]: Unless if I'm mistaken (and I could be :)), I don't think we should call these methods `Composite` as if it targets only one index, it's a regular index, not a composite one.

So let's rename them to `createIndex()`/`isIndexMissing()`.

File: neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/utils/Neo4jTestHelper.java [hunk 5]
```diff
@@ -137,6 +151,50 @@ private static DatastoreProvider getDatastoreProvider(SessionFactory sessionFact
...
+			List<Neo4jIndexSpec> indexes = toStream( neo4jDb.schema().getIndexes() )
+					.map( i -> new Neo4jIndexSpec( i.getLabel(), toList( i.getPropertyKeys() ), i.isConstraintIndex() ) )
+					.collect( Collectors.toList() );
+			tx.success();
+			return indexes;
+		}
+	}
+
+	private static List<Neo4jIndexSpec> getHttpNeo4jIndexes(DatastoreProvider provider) {
+		// todo: neo4jHttpClient does not return graph values
```
Comment (ntviet18) [inline neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/utils/Neo4jTestHelper.java]: @DavideD , I'm trying to get indexes by CALL db.indexes via HttpNeo4jClient. However it returns
```
{
  "results" : [ {
    "columns" : [ "description", "label", "properties", "state", "type", "provider" ],
    "data" : [ {
      "row" : null,
      "graph" : {
        "nodes" : [ ],
        "relationships" : [ ]
      },
      "meta" : null
    },... ],
  "errors" : [ ]
}
```

in contrast of direct rest response

```
{
    "results": [
        {
            "columns": [ "description",  "label", "properties", "state", "type", "provider" ],
            "data": [
                {
                    "row": [
                        "INDEX ON :Neo4jIndexTest$Person(age)",
                        "Neo4jIndexTest$Person",
                        [
                            "age"
                        ],
                        "ONLINE",
                        "node_label_property",
                        {
                            "version": "1.0",
                            "key": "lucene+native"
                        }
                    ]
                }, ...
    ],
    "errors": []
}
```

do we have any filter somewhere?

Comment (ntviet18) [reply neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/utils/Neo4jTestHelper.java]: Never mind, I have found my mistake. There are two modes for the query to work. I changed to `row` mode and everything works as expected.

**Comment by ntviet18** [comment]
@DavideD , @gsmet , could you please re-check? I have added some positive and negative test cases for all Neo4j provider types. It works locally for me with Neo4j v3.2.

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.Objects;
+
+import org.hibernate.ogm.datastore.neo4j.query.parsing.cypherdsl.impl.CypherDSL;
+
+import org.neo4j.graphdb.Label;
+
+/**
+ * @author The Viet Nguyen
+ */
+public class Neo4jIndexSpec {
```
Comment (fax4ever) [inline neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java]: could you add implements Serializable?

File: neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+import org.neo4j.graphdb.Label;
+
+import static org.fest.assertions.Assertions.assertThat;
+import static org.hibernate.ogm.datastore.neo4j.test.util.ExceptionHelper.extract;
+
+/**
+ * @author The Viet Nguyen
+ */
+@TestForIssue(jiraKey = "OGM-1462")
+public class Neo4jIndexTest extends Neo4jJpaTestCase {
```
Comment (fax4ever) [inline neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java]: Could you add a test where the index has applied on column, where the **column** name it is different from **field** name?

File: neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import org.neo4j.graphdb.Label;
+
+import static org.fest.assertions.Assertions.assertThat;
+import static org.hibernate.ogm.datastore.neo4j.test.util.ExceptionHelper.extract;
+
+/**
```
Comment (fax4ever) [inline neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java]: Could you add some text on JavaDoc expanding the **scope** of the test?

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/http/impl/HttpNeo4jSchemaDefiner.java [hunk 9]
```diff
@@ -53,6 +54,19 @@ protected void createSequences(List<Sequence> sequences, Set<IdSourceKeyMetadata
...
 
+	@Override
+	protected void createIndexesIfMissing(DatastoreProvider provider, List<Neo4jIndexSpec> indexes) {
+		Statements statements = new Statements();
+		for ( Neo4jIndexSpec index : indexes ) {
+			log.tracef( "Creating composite index for nodes labeled as %1$s on properties %2$s", index.getLabel(), index.getProperties() );
+			statements.addStatement( new Statement( index.asCypherQuery() ) );
+		}
+		log.debug( "Creating missing indexes" );
+		HttpNeo4jClient remoteClient = ( (HttpNeo4jDatastoreProvider) provider ).getClient();
```
Comment (fax4ever) [inline neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/http/impl/HttpNeo4jSchemaDefiner.java]: When we close the instance of HttpNeo4jClient?

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
+package org.hibernate.ogm.datastore.neo4j.impl;
```
Comment (fax4ever) [inline neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java]: Maybe you can move this class from *.neo4j.impl to *.neo4j.index.impl to be consistent with the same class of MongoDB?

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/BaseNeo4jSchemaDefiner.java [hunk 11]
```diff
@@ -82,6 +87,35 @@ public String toString() {
 
 	private static final Log log = LoggerFactory.make( MethodHandles.lookup() );
 
+	protected void createIndexes(DatastoreProvider provider, Database database) {
+		List<Neo4jIndexSpec> indexes = new ArrayList<>();
```
Comment (fax4ever) [inline neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/BaseNeo4jSchemaDefiner.java]: Maybe we can refactor the **indexes** local variable as a Class field, like in MongoDBSchemaDefiner class.
You could fill the list on validateMapping and create the indexes on initializeSchema, like in MongoDBSchemaDefiner. WDYT?

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java (hunk 10 above)
Comment (ntviet18) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java]: sure, I moved the tests but for got the main class 😃

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java (hunk 6 above)
Comment (ntviet18) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java]: sure, but may I ask why should we do that? We don't move this anywhere outside the JVM yet. And, I don't see `MongoDBIndexSpec` implements `Serialzable` either

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/http/impl/HttpNeo4jSchemaDefiner.java (hunk 9 above)
Comment (ntviet18) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/http/impl/HttpNeo4jSchemaDefiner.java]: I think it's when we close the `HttpNeo4jDatastoreProvider`

File: neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java (hunk 8 above)
Comment (ntviet18) [reply neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java]: sure

File: neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java (hunk 7 above)
Comment (ntviet18) [reply neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java]: interesting case, I sure will add it

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java (hunk 6 above)
Comment (fax4ever) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java]: It's an optional, sometimes it is better to mark a class Serializable, it a way to describe the nature of the class itself.

File: neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java (hunk 7 above)
Comment (fax4ever) [reply neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java]: Thank you

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/http/impl/HttpNeo4jSchemaDefiner.java (hunk 9 above)
Comment (fax4ever) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/http/impl/HttpNeo4jSchemaDefiner.java]: you're right :P

File: neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java (hunk 6 above)
Comment (ntviet18) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java]: I get it. Will add in the next push

Comment (fax4ever) [reply neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/Neo4jIndexSpec.java]: Now we are more focus on what is Serializable or not, more than in the past, for cloud compliance :P

File: neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java (hunk 7 above)
Comment (ntviet18) [reply neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/index/Neo4jIndexTest.java]: I have checked, this case is already covered by hibernate-core

**Comment by ntviet18** [comment]
Hello everyone, could you please tell me what is the version we are using on CI?

**Comment by DavideD** [comment]
> Hello everyone, could you please tell me what is the version we are using on CI?

Hi, I'm upgrading to 3.3.5 but I'm having some issues with the buid right now. Working on it. If you see some failure, it's likely not because of this PR.

**Comment by DavideD** [comment]
Thanks a lot @ntviet18 ,
I'm still finishing the review but overall it looks good.

The only real problem I have with this PR is that you squashed all changes in a single commit. We always try to send PR as a collection of commits containing logical changes. So far I can see at least 3 commits: one for the creation of the ExceptionHjelper class,  one for the fix and one for the changes related to the tests used to check that the indexes actually exist no the datastore.

Having these changes separated has several advantages:
* it makes it easier to understand what the changes are about and therefore it's easier to review them;
* if the tests are in a separate commit I can run them on master and verify that they really fail without the fix;
* if a group of changes are not right or not exactly what we need I can merge the rest of the fix quite easily meaning that it takes me less time to merge a PR, or just update the commit that is not right;
* having new classes in separate commits (if possible) helps to highlight what's the new stuff that has been introduced and make it easier to rebase when needed;
* if I want to apply some of the changes to a different branch I can just cherry-pick the related commits;
* if we realize that something is not right in the future we might be able to simply revert the commit invoved;
* we might decide that some changes can be promoted to a new JIRA;
* it's easier to squashed commits together if needed than split an existing commit.

We still want the project to build for every commit and this means that sometimes it's not possible to completely keep the changes separated (for example, tests cannot be kept separated or need to be included after the fix). But we still try.

If it's not clear what I mean, I would suggest to have a look at the commit log. You can also check the other PR in the https://github.com/hibernate/hibernate-search or https://github.com/hibernate/hibernate-validator repositories. 

***No need to change this PR now*** but please keep this in mind for the next PR.

Thanks a lot again for the great work.

I will now continue the review ...

**Comment by ntviet18** [comment]
@DavideD , sorry about that. I used a custom git-review tool that lazily squash all commits together. I will break commit to parts in the future.

**Comment by DavideD** [comment]
Just a note to clarify that I haven't forgotten this PR. I'm having some problems upgrading Neo4j to version 3.3.5 and I'm working on it.

**Comment by DavideD** [comment]
This PR requires some changes, I'm going to close it for now and send a new one soon build on top of this one.

Cheers

**Pull request closed without merging**